File: design-patterns/java/templatemethod/PdfReportExporter.java
import java.util.Map;

public class PdfReportExporter extends AbstractReportExporter {
    //prepareData() not overridden - default will be used
    //openFile() not overridden - default will be used

    @Override
    protected void writeHeader(ReportData data) {
        System.out.println("PDF: Writing header: " + String.join(",", data.getHeaders()));
    }

    @Override
    protected void writeDataRows(ReportData data) {
        System.out.println("PDF: Writing data rows...");
        for (Map<String, Object> row : data.getRows()) {
            System.out.println("PDF: " + row.values());
        }
    }

    // writeFooter() not overridden - default will be used
    // closeFile() not overridden - default will be used
}

File: design-patterns/cpp/templatemethod/pdf_report_exporter.cpp
#include "pdf_report_exporter.h"
#include <iostream>

void PdfReportExporter::writeHeader(const ReportData& data) {
    std::cout << "Writing PDF header..." << std::endl;
    std::cout << "=== Report Header ===" << std::endl;
    for (const auto& header : data.getHeaders()) {
        std::cout << header << "\t";
    }
    std::cout << std::endl;
    std::cout << "===================" << std::endl;
}

void PdfReportExporter::writeDataRows(const ReportData& data) {
    std::cout << "Writing PDF data rows..." << std::endl;
    for (const auto& row : data.getRows()) {
        for (const auto& header : data.getHeaders()) {
            const auto& value = row.at(header);
            std::visit([](const auto& v) { std::cout << v; }, value);
            std::cout << "\t";
        }
        std::cout << std::endl;
    }
} 